atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -0,0 +1,227 @@
+"""
+Copyright (c) 2019 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+from __future__ import absolute_import
+
+import os
+import tempfile
+import time
+
+import requests
+from six.moves.urllib.parse import urlparse
+
+from atomic_reactor.constants import (
+    DEFAULT_DOWNLOAD_BLOCK_SIZE,
+    HTTP_BACKOFF_FACTOR,
+    HTTP_MAX_RETRIES,
+    PLUGIN_FETCH_SOURCES_KEY
+)
+from atomic_reactor.plugin import PreBuildPlugin
+from atomic_reactor.plugins.pre_reactor_config import (
+    get_koji,
+    get_koji_path_info,
+    get_koji_session,
+    get_config,
+    NO_FALLBACK
+)
+from atomic_reactor.util import get_retrying_requests_session
+
+
+class FetchSourcesPlugin(PreBuildPlugin):
+    """Download sources that may be used in further steps to compose Source Containers"""
+    key = PLUGIN_FETCH_SOURCES_KEY
+    is_allowed_to_fail = False
+    DOWNLOAD_DIR = 'image_sources'
+
+    def __init__(
+        self, tasker, workflow, koji_build_id=None, koji_build_nvr=None, signing_intent=None,
+    ):
+        """
+        :param tasker: ContainerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param koji_build_id: int, container image koji build id
+        :param koji_build_nvr: str, container image koji build NVR
+        :param signing_intent: str, ODCS signing intent name
+        """
+        if not koji_build_id and not koji_build_nvr:
```

Comment:
and what if both are provided?
in set_koji_image_build_data
self.koji_build_nvr is favored 

but I think we should raise error if both are provided

Reply:
That's what we are doing in `set_koji_image_build_data`: if both are provided, they should refer to the same build (we are checking this through the `build_id`).

Reply:
right I see it now